operator: Wait for CRDs before running informers

vadorovsky opened the pull request:
Running CRD informers was sometimes resulting in errors when
cilium-operator was launched before cilium-agent created CRDs. This
change adds the waitForCRD() function which waits for the given CRD with
the timeout.

Fixes: #10814

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

```release-note
cilium-operator: Wait for CRDs before running Informers
```

vadorovsky committed:
operator: Wait for CRDs before running informers

Running CRD informers was sometimes resulting in errors when
cilium-operator was launched before cilium-agent created CRDs. This
change adds the waitForCRD() function which waits for the given CRD with
the timeout.

Fixes: #10814

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky committed:
vendor: Add missing modules from apiextensions-apiserver and client-go

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,55 @@
+}
+
+func getCRD(client clientset.Interface, CRDName string) error {
+	_, err := client.ApiextensionsV1beta1().CustomResourceDefinitions().Get(context.TODO(), CRDName, metav1.GetOptions{})
```
it needs the following in the clusterrole of cilium-operator

```
- apiGroups:
  - apiextensions.k8s.io
  resources:
  - customresourcedefinitions
  verbs:
  - get
```

vadorovsky replied:
Thanks! I was testing it on vagrant, where operator is managed by systemd... I should've try minikube and microk8s as well.

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,55 @@
+	}); err != nil {
+		return errors.Wrapf(err, "timeout exceeded when waiting for CRD %s", CRDname)
+	}
+	return nil
```
couldn't this be replaced with a cilium controller? It will also retry on error and we don't need to import more external packages.

vadorovsky replied:
Cilium controller seems to have much more complex purpose than just simple "retry on error" action which I need here? I can try to think of the way of reusing parts of it, or rewriting it a bit in a way it will be easy to use for CRD watching.

vadorovsky replied:
Or nevermind, sorry, I think I can use pkg/controller without any changes.

vadorovsky committed:
helm: Allow to watch CRDs by operator

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

christarazi commented on operator/main.go:
```diff
@@ -194,7 +205,9 @@ func runOperator(cmd *cobra.Command) {
-		startSynchronizingCiliumNodes(nodeManager)
+		if err := startSynchronizingCiliumNodes(nodeManager); err != nil {
+			log.WithError(err).Fatal("Unable to start synchronizing Cilium nodes")
+		}
```
The code in this switch statement is basically the same except for the cloud provider. We can eliminate the duplication by extracting the minimal necessary information from the switch statement and then proceeding outside the switch statement with the generic operation, like calling `Init()`, and the rest. 

For example, it seems like the only piece of info we need out of this is the specific allocator and provider.

```
ipamAllocatorAWS, providerBuiltin := allocatorProviders["aws"]
```
```
ipamAllocatorAzure, providerBuiltin := allocatorProviders["azure"] 
```

vadorovsky replied:
The code has changed during the last rebase, everything is handled in the one case now,

christarazi commented on operator/crd.go:
```diff
@@ -0,0 +1,54 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func getCRD(client clientset.Interface, CRDName string) error {
```
Nit: `CRDName` can just be `name` or `crdName`.

vadorovsky replied:
done

christarazi commented on operator/crd.go:
```diff
@@ -0,0 +1,54 @@
+
+// waitForCRD waits for the given CRD to be available until the timeout defined
+// by CRDRetry. Returns an error when timeout exceeded.
+func waitForCRD(client clientset.Interface, CRDname string) error {
```
Nit: `CRDName` can just be `name` or `crdName`.

vadorovsky replied:
done

christarazi commented on operator/crd.go:
```diff
@@ -0,0 +1,54 @@
+// by CRDRetry. Returns an error when timeout exceeded.
+func waitForCRD(client clientset.Interface, CRDname string) error {
+	ctrlMgr := controller.NewManager()
+	ctrlName := fmt.Sprintf("wait-for-crd-%s", CRDname)
```
I think this can just be a simple concatenation instead of a call to `Sprintf`. Helps remove an import of `fmt` :smile:

vadorovsky replied:
done

christarazi reviewed (changes requested):
Thanks for the PR! Looks good, aside for a few minor comments

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,53 @@
+		return errors.Wrapf(err, "Could not wait for CRD %s status", name)
+	}
+	return nil
+}
```
I'm a little lost by this function. It seems it will create a controller and then will remove it. Is this intended?

vadorovsky replied:
Yes - my intention is to initially wait until we can get the CRD and then remove the contoller once we get it successfully. I don't see any need to have that controller running afterwards.

aanm replied:
Right but line 43 is non blocking, i.e. the controller will be updated and removed immediately.

vadorovsky replied:
`RemoveControllerAndWait` will wait until the controller has completed and as far as I understand, it's blocking. That's my log from cilium-operator:

```
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Getting CRD ciliumnetworkpolicies.cilium.io from Kubernetes apiserver" subsys=cilium-operator
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Controller func execution time: 1.336µs" name=kvstore-sync-store-cilium/state/cnpstatuses/v2 subsys=controller uuid=5a5870de-991e-11ea-820a-08002737a2e9
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Controller run succeeded; waiting for next controller update or stop" name=kvstore-sync-store-cilium/state/cnpstatuses/v2 subsys=controller uuid=5a5870de-991e-11ea-820a-08002737a2e9
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Starting watcher..." config=/var/lib/cilium/etcd-config.yml endpoints="[https://192.168.33.11:2379]" subsys=kvstore watcher=cnpStatusWatcher
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Distributed lock successful, etcd has quorum" config=/var/lib/cilium/etcd-config.yml endpoints="[https://192.168.33.11:2379]" subsys=kvstore
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="List response from etcd len=0: &{Header:cluster_id:6438720328480595214 member_id:1326427008500738502 revision:408 raft_term:2  Kvs:[] More:false Count:0}" config=/var/lib/cilium/etcd-config.yml endpoints="[https://192.168.33.11:2379]" prefix=cilium/state/cnpstatuses/v2 subsys=kvstore watcher=cnpStatusWatcher
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Starting to watch a prefix" config=/var/lib/cilium/etcd-config.yml endpoints="[https://192.168.33.11:2379]" prefix=cilium/state/cnpstatuses/v2 revision=409 subsys=kvstore watcher=cnpStatusWatcher
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Check of CRD ciliumnetworkpolicies.cilium.io successful" subsys=cilium-operator
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Controller func execution time: 55.037927ms" name=wait-for-crd-ciliumnetworkpolicies.cilium.io subsys=controller uuid=5a58728f-991e-11ea-820a-08002737a2e9
May 18 15:43:43 k8s1 cilium-operator[18801]: level=debug msg="Shutting down controller" name=wait-for-crd-ciliumnetworkpolicies.cilium.io subsys=controller uuid=5a58728f-991e-11ea-820a-08002737a2e9
```

Which seems to prove that the controller was blocked until it finished its successful iteration.

But on the other hand, it's the first time when I ever use the controller, so I might be wrong.

vadorovsky replied:
Anyway, if I'm wrong, I'm open for suggestions how should I use the controller to wait until it's first successful run.

vadorovsky replied:
Or no, sorry, I'm wrong. There is no guarantee that there was at least one successful iteration of controller when using `RemoveControllerAndWait`. I need to come up with some solution which will ensure at least one success.

vadorovsky replied:
Should be fixed now by 62ce6814d6429b79d2ce0a351a123f28bac68124

aanm commented on operator/k8s_identity.go:
```diff
@@ -169,5 +169,9 @@ func startManagingK8sIdentities() {
 		identityStore,
 	)
 
+	if err := waitForCRD(apiextensionsK8sClient, "ciliumidentities.cilium.io"); err != nil {
```
Isn't `ciliumidentities` + `.cilium.io` defined anywhere?

vadorovsky replied:
`ciliumidentities` is not defined anywhere as a global variable, it's defined as a local variable here:

https://github.com/cilium/cilium/blob/3a72ffc4f1808c29077d5ebb645c1b28af257996/pkg/k8s/apis/cilium.io/v2/client/register.go#L463-L479

But defining CRD names as global (const) variables is a good idea, I can do that.

vadorovsky replied:
Seems like `pkg/k8s/apis/cilium.io/const.go` is a good place for global variables. I will move them there.

aanm commented on operator/cnp_event.go:
```diff
@@ -136,6 +136,10 @@ func enableCNPWatcher() error {
 		cnpStore,
 	)
+
+	if err := waitForCRD(apiextensionsK8sClient, "ciliumnetworkpolicies.cilium.io"); err != nil {
```
Isn't `ciliumnetworkpolicies` + `.cilium.io` defined anywhere?

vadorovsky replied:
ditto

aanm commented on operator/cilium_node.go:
```diff
@@ -72,12 +72,17 @@ func startSynchronizingCiliumNodes(nodeManager allocator.NodeEventHandler) {
 		k8s.ConvertToCiliumNode,
 	)
 
+	if err := waitForCRD(apiextensionsK8sClient, "ciliumnodes.cilium.io"); err != nil {
```
Isn't `ciliumnodes` + `.cilium.io` defined anywhere?

vadorovsky replied:
ditto

vadorovsky committed:
k8s: Define CRD names and parameters as global const variables

Before this change, variables for CRD names and parameters were local
and could not be used outside the package.

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

vadorovsky commented on pkg/controller/controller.go:
```diff
@@ -259,18 +264,42 @@ func (c *Controller) runController() {
 			select {
 			case <-c.stop:
 				goto shutdown
+			case <-c.stopOnSuccessCh:
```
I don't like the fact that I had to copy the whole `case <-c.stopOnSuccessCh` section here, but I had no better idea. I'm open for suggestions.

christarazi replied:
One simpler approach would be to put the `case` body into a helper function. I'm not sure what would happen to the `goto` statements, but you'll likely need to refactor the `goto`s out in order to use this approach. Just an idea. :)

christarazi replied:
If I understand correctly, the code written here depends on reading from channels in some order. Golang's `select` statement randomness is causing you to nest the `select` statement and duplicate the case body into another case, because you can't know in which order the channels became ready.

Another approach could be to separate out the `select` statement for only the use of reading from `c.stopOnSuccessCh` (assuming that's the one you want to give priority to). Then you could keep the other `select` with `c.update`'s case body slimmer (without the duplicated `c.stopOnSuccessCh` case) in there. Does that make sense? Happy to clarify :)

vadorovsky replied:
Thanks for suggestions! :) By following @aanm's suggestion, I don't need nay changes in controller anymore.

aanm commented on pkg/k8s/apis/cilium.io/register.go:
```diff
@@ -15,6 +15,86 @@
+	CustomResourceDefinitionSchemaVersion = "1.20"
+
+	// CustomResourceDefinitionSchemaVersionKey is key to label which holds the CRD schema version
+	CustomResourceDefinitionSchemaVersionKey = "io.cilium.k8s.crd.schema.version"
```
At list these consts are bounded to a particular version so they should belong to at least `...cilium.io/v2/register.go`

vadorovsky replied:
Done.

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,62 @@
+// by CRDRetry. Returns an error when timeout exceeded.
+func WaitForCRD(client clientset.Interface, name string) error {
+	return waitForCRD(client, name, defaultRunInterval, defaultTimeout)
+}
```
Sorry, I gave a little more thought to this section. Instead of poling we could do something similar to this https://github.com/kubernetes/kubernetes/pull/88935/commits/e053fdd08ae00d1bff3dd42a2c034f4810f8a55d#diff-934760ea1a4e195e49bcd10625cf4e46R384 where we would watch and wait until that particular CRD is populated. This wouldn't require any changes in the controller code.

vadorovsky replied:
Good idea! Done.

aanm commented on install/kubernetes/cilium/charts/operator/templates/clusterrole.yaml:
```diff
@@ -53,3 +53,9 @@ rules:
+  resources:
+  - customresourcedefinitions
+  verbs:
+  - get
```
Based on the suggestions we would need to add `list` and `watch` verbs here.

vadorovsky replied:
Done

aanm commented on install/kubernetes/quick-install.yaml:
```diff
@@ -317,6 +317,12 @@ rules:
+  resources:
+  - customresourcedefinitions
+  verbs:
+  - get
```
Based on the suggestions we would need to add `list` and `watch` verbs here.

vadorovsky replied:
Done

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,83 @@
+	defer cancelFunc()
+
+	selector := fields.OneTermEqualSelector("metadata.name", name).String()
+	fmt.Println(selector)
```
leftover

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,83 @@
+		return errors.Wrapf(err, "timeout waiting for CRD %s", name)
+	}
+	if crd, ok := ev.Object.(*v1beta1.CustomResourceDefinition); ok {
+		fmt.Println(crd)
```
leftover

aanm commented on pkg/policy/groups/helpers.go:
```diff
@@ -20,10 +20,9 @@ import (
 	"github.com/cilium/cilium/pkg/policy/api"
 
-	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
unnecessary change

vadorovsky replied:
I don't know why, but my gofmt/golint hook always adds that. :( And I've seen those `v1 "k8s.io/foo/v1"` imports in many places and even in Cilium. I can't google why this is happening. I guess I have to disable lint hook on save for now.

vadorovsky replied:
OK, that was because of `goimports`. I will stop using it for Cilium.

vadorovsky replied:
Fixed

aanm commented on operator/k8s_identity.go:
```diff
@@ -119,13 +119,13 @@ func startCRDIdentityGC() {
 }
 
-func startManagingK8sIdentities() {
+func startManagingK8sIdentities() error {
```
can you pass `apiextensionsK8sClient` as an argument of this function and add documentation stating the function will wait until the creation of the CRD for `CiliumIdentities`?

vadorovsky replied:
~Why should I pass a global variable as an argument? `ciliumK8sClient` is a global variable too and it's used straight in `k8s_identity.go`, so I just wanted to be consistent. Keeping k8s clients as globals seems to be an approach used in the whole Cilium. Having one client (`ciliumK8sClient`) as a global, and the other client (`apiextensionsK8sClient`) as a local variable in `main` being passed seems a bit odd to me.~

~I agree about adding a documentation/comment to the function.~

aanm is right, I will fix that

vadorovsky replied:
done

aanm commented on operator/k8s_identity.go:
```diff
@@ -22,12 +22,12 @@ import (
 	"github.com/cilium/cilium/operator/watchers"
 	"github.com/cilium/cilium/pkg/controller"
-	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
```
unnecessary change

aanm commented on operator/k8s_identity.go:
```diff
@@ -22,12 +22,12 @@ import (
 
 	"github.com/sirupsen/logrus"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
unnecessary change

aanm commented on operator/errors.go:
```diff
@@ -0,0 +1,23 @@
+package main
+
+import (
+	"github.com/pkg/errors"
```
`s+github.com/pkg/errors+errors`

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,80 @@
+
+func waitForCRD(client clientset.Interface, name string, timeout time.Duration) error {
+	ctx, cancelFunc := context.WithTimeout(context.Background(), timeout)
+	defer cancelFunc()
```
Instead of passing `timeout time.Duration` can't the function receive the context instead?

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,80 @@
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
```
`s+github.com/pkg/errors+errors`

aanm commented on operator/cnp_event.go:
```diff
@@ -27,7 +27,7 @@ import (
 	"github.com/cilium/cilium/pkg/policy/groups"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
unnecessary change

aanm commented on operator/cnp_event.go:
```diff
@@ -27,7 +27,7 @@ import (
+	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/util/runtime"
 	"k8s.io/apimachinery/pkg/util/wait"
```
line49: can you pass `apiextensionsK8sClient` as an argument of this function and add documentation stating the function will wait until the creation of the CRD for `CiliumNetworkPolicies`?

vadorovsky replied:
done

aanm commented on operator/cilium_node.go:
```diff
@@ -19,11 +19,11 @@ import (
 	k8sversion "github.com/cilium/cilium/pkg/k8s/version"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
unnecessary change

aanm commented on operator/cilium_node.go:
```diff
@@ -19,11 +19,11 @@ import (
 	"github.com/cilium/cilium/pkg/ipam/allocator"
 	"github.com/cilium/cilium/pkg/k8s"
-	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
```
unnecessary change

aanm commented on operator/cilium_node.go:
```diff
@@ -32,15 +32,15 @@ import (
 var k8sCiliumNodesCacheSynced = make(chan struct{})
 
-func startSynchronizingCiliumNodes(nodeManager allocator.NodeEventHandler) {
+func startSynchronizingCiliumNodes(nodeManager allocator.NodeEventHandler) error {
```
can you pass `apiextensionsK8sClient` as an argument of this function and add documentation stating the function will wait until the creation of the CRD for `CiliumNodes`?

vadorovsky replied:
done

aanm commented on operator/crd.go:
```diff
@@ -0,0 +1,80 @@
+)
+
+const (
+	defaultTimeout = 300 * time.Second
```
Any reason for the 5 minutes? I think this can be decreased to `2 * time.Minute`.

aanm reviewed (approved):
:100:

christarazi reviewed (approved):
Well done!

joestringer reviewed (approved):
Policy / helm changes LGTM.

joestringer commented on operator/crd.go:
```diff
@@ -0,0 +1,82 @@
+
+// WaitForCRD waits for the given CRD to be available until the default timeout,
+// after which cilium-agent should be ready. Returns an error when timeut
+// esceeded.
```
```suggestion
// after which cilium-agent should be ready. Returns an error when timeout
// is exceeded.
```

vadorovsky replied:
Done

joestringer commented on operator/crd_test.go:
```diff
@@ -0,0 +1,76 @@
+	c.Assert(err, IsNil)
+
+	// Try to get non-existing CRD
+	err = waitForCRD(context.TODO(), client, "bar")
+	c.Assert(err, ErrorMatches, ".*timeout waiting for CRD bar.*")
```
What's the timeout here? Seems like it should be the one coming from `UntilWithSync()` but I can't tell what it would be by default. Mainly just want to make sure we're not just adding arbitrary idle time into the unit test runs.

On a related note, I don't see this test name in the Travis output and when I tried to run it locally, it didn't get automatically hooked into the standard CI targets:

```
$ SKIP_VET=1 SKIP_KVSTORES=1 make unit-tests TESTPKGS=operator
dirname: missing operand
Try 'dirname --help' for more information.
  GO    tools/maptool/maptool
dirname: missing operand
Try 'dirname --help' for more information.
=== RUN   Test
OK: 1 passed
--- PASS: Test (0.00s)
=== RUN   Test_getOldestLeases
=== RUN   Test_getOldestLeases/test-1
=== RUN   Test_getOldestLeases/test-2
--- PASS: Test_getOldestLeases (0.00s)
    --- PASS: Test_getOldestLeases/test-1 (0.00s)
    --- PASS: Test_getOldestLeases/test-2 (0.00s)
=== RUN   Test_getPath
=== RUN   Test_getPath/test-1
=== RUN   Test_getPath/test-2
--- PASS: Test_getPath (0.00s)
    --- PASS: Test_getPath/test-1 (0.00s)
    --- PASS: Test_getPath/test-2 (0.00s)
PASS
coverage: 17.1% of statements in github.com/cilium/cilium/operator
ok      github.com/cilium/cilium/operator       0.037s  coverage: 17.1% of statements in github.com/cilium/cilium/operator
dirname: missing operand
Try 'dirname --help' for more information.
dirname: missing operand
Try 'dirname --help' for more information.
```

joestringer replied:
The above assumption I made is wrong, apparently it is run, it's just that gocheck doesn't report it for whatever reason (maybe I'm missing gocheck-specific verbosity flags):

```
$ git diff
diff --git a/operator/crd_test.go b/operator/crd_test.go
index 81d4435e9a89..75b4c70229ba 100644
--- a/operator/crd_test.go
+++ b/operator/crd_test.go
@@ -73,4 +73,6 @@ func (s *crdTestSuite) TestGetCRD(c *C) {
        // Try to get non-existing CRD
        err = waitForCRD(context.TODO(), client, "bar")
        c.Assert(err, ErrorMatches, ".*timeout waiting for CRD bar.*")
+
+       c.Assert(true, Equals, false)
 }
$ SKIP_VET=1 SKIP_KVSTORES=1 make unit-tests TESTPKGS=operator
dirname: missing operand
Try 'dirname --help' for more information.
  GO    tools/maptool/maptool
dirname: missing operand
Try 'dirname --help' for more information.
=== RUN   Test

----------------------------------------------------------------------
FAIL: crd_test.go:63: crdTestSuite.TestGetCRD

crd_test.go:77:
    c.Assert(true, Equals, false)
... obtained bool = true
... expected bool = false

OOPS: 0 passed, 1 FAILED
--- FAIL: Test (0.00s)
=== RUN   Test_getOldestLeases
=== RUN   Test_getOldestLeases/test-1
=== RUN   Test_getOldestLeases/test-2
--- PASS: Test_getOldestLeases (0.00s)
    --- PASS: Test_getOldestLeases/test-1 (0.00s)
    --- PASS: Test_getOldestLeases/test-2 (0.00s)
=== RUN   Test_getPath
=== RUN   Test_getPath/test-1
=== RUN   Test_getPath/test-2
--- PASS: Test_getPath (0.00s)
    --- PASS: Test_getPath/test-1 (0.00s)
    --- PASS: Test_getPath/test-2 (0.00s)
FAIL
coverage: 17.1% of statements in github.com/cilium/cilium/operator
FAIL    github.com/cilium/cilium/operator       0.039s
FAIL
make: *** [Makefile:225: unit-tests] Error 1
```

joestringer commented on operator/crd.go:
```diff
@@ -0,0 +1,82 @@
+)
+
+const (
+	defaultTimeout = 2 * time.Minute
```
Followup: would be nice to be able to configure this in case it lines up poorly with any other timeouts.

vadorovsky replied:
Will do in a separate PR.

joestringer reviewed (approved):
I think the core change here looks right; there's definitely some points for follow up noted below. 

EDIT: Given that my unit-test concern is mitigated, these are more "nice-to-haves".
